### PR TITLE
Morph Ambushing (Code Improvements, Balance Changes, Yadda Yadda.)

### PR DESCRIPTION
**Pull request by tralezab**
## About The Pull Request

Mostly loyal to https://tgstation13.org/phpBB/viewtopic.php?f=9&t=29093&view=unread#unread <this thread, though I haven't gone as hard as they did with damage increases.

Morphs can no longer be AI, they're really bad and this was a bad idea and we still feel the pain and I hate it and now it's gone!

You no longer are unable to attack while disguised, but undisguise when you do. Morphs now normally have a low attack that is high while disguised.

Morphs are now ambushers. After assuming the form of objects, they will prepare special attacks after 5 seconds of standing still.

Special attacks depend on what you disguise as, but I've left this pretty wide open until someone suggests some neat ones. Maybe i'll do special ambushes for guns too

For most items, a hand attack will trigger when interacted with that dismembers the hand that interacted and deals massive damage.
For floor items like vents and scrubbers, a floor attack will trigger when stepped on that dismembers legs and deals massive damage.

Like... A lot of other minor balance stuffs. Sorry I probably missed something.

- [ ] add in that sound for morph ambushing
- [ ] cooldown for redisguising
- [ ] some more ambushes
- [ ] morphs copy the access of id cards they consume

## Code Improvements

Morph spawn event is now in its own file in events
No more single letter vars.
Removed a bunch of terrible vars on the morph itself.

- [ ] doc it all

## Why It's Good For The Game

Morphs are shit, I read the forum thread and this seemed like... better than we have now. Maybe not by a mile, but enough to make it worth doing.

## Changelog
:cl:
balance: Morphs have gotten a moderate gameplay rework. They now set up ambushes as disguised items while still.
balance: The special attacks they do are dependent on what they disguise as. Try disguising as a floor vent!
code: Oh, and the code is better
/:cl:

### CODE REVIEW

**Comment by tralezab** [comment]
This broke my master but luckily for me that is a problem for future me

**Comment by ilikesaying** [comment]
would ambush damage the dismembered limb or a random bodypart that's not the limb?

**Comment by D4C-420** [comment]
If we buffing morph why not make it a Mid-round same as swarmer's or a space dragon ? Currently aren't miners the only ones who become morphs outside of trades?

**Comment by tralezab** [comment]
@D4C-420 I'll keep enabling morph to a separate pr

@ilikesaying the damage is run post dismember, so it is applied everywhere else.

File: code/modules/antagonists/morph/morph.dm [hunk 1]
```diff
@@ -1,129 +1,239 @@
...
+#define AMBUSH_INTERACT (1>>0)
+///ambush on someone walking over you (vents)
+#define AMBUSH_WALKED_OVER (1>>1)
+
+///damage applied to someone for a successful ambush
+#define AMBUSH_DAMAGE 70
+///range at which morph disguises show an odd examine message
+#define MORPH_EXAMINE_REVEAL_DISTANCE 3
+///the time the morph needs to sit still for the ambush to activate!
+#define TIME_TO_AMBUSH 5 SECONDS
```
Comment (tralezab) [inline code/modules/antagonists/morph/morph.dm:12]: need to be undefined... or maybe moved into a _DEFINES file

**Comment by IndieanaJones** [comment]
I don't think this really solves the main issue with Morph's design, which is having a stealth antag based around consuming bodies.  Unlike changeling, who leaves their husks around which can have a few things done with them to bring the person back into the round, Morphs carry their victims around and they will only escape should they die, which means getting caught off-guard by a Morph ends up being a pretty lame way to die.  Here were my thoughts on reworking Morph:

- No longer based around murdering and consuming people via simplemob unga, but would be based around ambushing like the version you have here.  Instead of going for the kill, the Morph would have a lunge which would undo the disguise and would pounce on the target, stunning them.
- Upon stunning a target, the Morph would take some organ, perhaps random or maybe something specific (I was thinking eyes, along with a tagline "Use em' or lose em'), and this would form the Morph's main objective.
- If the Morph bags so many whatever its after, something happens

I know its not detailed, but I did intend to re-work Morph at some point, not that I mind you doing it.  I just think fixing that first issue is of some importance.

**Comment by panvxv** [comment]
> 
> 
> I don't think this really solves the main issue with Morph's design, which is having a stealth antag based around consuming bodies. Unlike changeling, who leaves their husks around which can have a few things done with them to bring the person back into the round, Morphs carry their victims around and they will only escape should they die, which means getting caught off-guard by a Morph ends up being a pretty lame way to die. Here were my thoughts on reworking Morph:
> 
>     * No longer based around murdering and consuming people via simplemob unga, but would be based around ambushing like the version you have here.  Instead of going for the kill, the Morph would have a lunge which would undo the disguise and would pounce on the target, stunning them.
> 
>     * Upon stunning a target, the Morph would take some organ, perhaps random or maybe something specific (I was thinking eyes, along with a tagline "Use em' or lose em'), and this would form the Morph's main objective.
> 
>     * If the Morph bags so many whatever its after, something happens
> 
> 
> I know its not detailed, but I did intend to re-work Morph at some point, not that I mind you doing it. I just think fixing that first issue is of some importance. This concept would also do away with the consumption of random items as well, as losing something important to a morph can be just as annoying as dying and having your corpse be essentially gone. My plan was to delegate item-stealing to a new non-stealth antag who would built around the concept.

Problem with leaving is that people will now Morph is on station and now they will be more suspicious about items they pick up.
Stunning mechanic would be killer for Morphs when geneticists release Hulks.

**Comment by panvxv** [comment]
I don't know about the damage nerf to 10 damage. One mistake and you are dead to an assistant. At least give some penetration as ambushing sec officers wouldn't be good for you without it.
I know that damages would be high but this is not a normal traitor that can pick up fights. If it doesn't crit someone in 2 hits (ambush and attack) prey will just escape and inform others. With this traitor mistakes of being killed by it is on defender side.
You didn't check object, you tried to pick it up and you died.

**Comment by tralezab** [comment]
I'm ok changing around some numbers.

**Comment by techno0** [comment]
Have you considered removing the ability of the morph to eat all materials and maybe limit it to organic matter only? I don't really see a reason why you should be able to consume literally tons of metal,glass,plasma that the morph just drops after dying. I don't care about it eating guns or small stuff but the morph having the ability to just consume a planet worth of resources seems kinda dumb.

**Comment by tralezab** [comment]
I think it has become iconic of the morph but I would prefer if it actually meant something to eat what it does

**Comment by panvxv** [comment]
I have just written on my thread the idea of collecting items to fulfill goal. I would really appreciate any criticism especially from you tralezab.

**Comment by the-orange-cow** [comment]
> * If the Morph bags so many whatever its after, something happens

I like the idea of there being tangible benefit to the morph eating a lot of stuff.  Yeah, voring the entire armory/ORM is fun just for the sake of it, but it would be nice if there was real impetus for doing so.

**Comment by tralezab** [comment]
After testing, this was an improvement, but it requires a lot of polish and balance tweaks and i'm not interested enough to go down that very long road

**Pull request closed without merging**